util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+r"""OTP memory image class, used to create preload images for the OTP
+memory for simulations and FPGA emulation.
+"""
+
+import logging as log
+import random
+from collections import OrderedDict
+
+from lib.common import check_bool, check_int, ecc_encode, random_or_hexvalue
+from lib.LcStEnc import LcStEnc
+from lib.OtpMemMap import OtpMemMap
+from lib.Present import Present
+
+# Seed diversification constant for OtpMemImg (this enables to use
+# the same seed for different classes)
+OTP_IMG_SEED_DIVERSIFIER = 1941661965323525198146
+
+
+def _int_to_byte_array(value, num_bytes):
+    '''Convert arbitrary precision integer to raw byte string'''
+    value_hex = format(value, '0' + str(num_bytes * 2) + 'X')
+    return bytearray.fromhex(value_hex)
+
+
+def _byte_array_to_int(byte_str):
+    '''Convert raw byte string into arbitrary precision integer'''
+    value_hex = bytearray.hex(byte_str)
+    return int(value_hex, 16)
+
+
+def _present_64bit_encrypt(plain, key):
+    '''Scramble a 64bit block with PRESENT cipher'''
+
+    # Make sure data is within 64bit range
+    if (plain < 0) or (plain > 2**64):
+        raise RuntimeError('Data block is out of 64bit range')
+
+    # Make sure key is within 128bit range
+    if (key < 0) or (key > 2**128):
+        raise RuntimeError('Key is out of 128bit range')
+
+    # Make sure inputs are integers
+    if not isinstance(plain, int) or not isinstance(key, int):
+        raise RuntimeError('Data and key need to be of type int')
+
+    cipher = Present(_int_to_byte_array(key, 16))
+    encr = cipher.encrypt(_int_to_byte_array(plain, 8))
+    return _byte_array_to_int(encr)
+
+
+def _present_64bit_digest(data_blocks, iv, const):
+    '''Compute digest over multiple 64bit data blocks'''
+
+    # We need to align the number of data blocks to 2x64bit
+    # for the digest to work properly.
+    if len(data_blocks) % 2 == 1:
+        data_blocks.append(data_blocks[-1])
```

Comment:
This is a slight trap to the caller it mutates the argument. I'd probably suggest something like `data_blocks = data_blocks + [data_blocks[-1]]` to avoid it.

Reply:
fixed!